eZ/Publish/Core/Persistence/Cache/UserHandler.php

```diff
@@ -495,6 +497,94 @@ function () use ($roleId) {
+    {
+        return $this->getListCacheValue(
+            $this->cacheIdentifierGenerator->generateKey(
+                self::ROLE_ASSIGNMENT_WITH_BY_ROLE_SUFFIX_OFFSET_LIMIT_IDENTIFIER,
+                [$roleId, $offset, $limit],
+                true
+            ),
+            function () use ($roleId, $offset, $limit): array {
+                return $this->persistenceHandler
+                    ->userHandler()
+                    ->loadRoleAssignmentsByRoleIdWithOffsetAndLimit($roleId, $offset, $limit);
+            },
+            $this->getRoleAssignmentTags,
+            $this->getRoleAssignmentKeys,
+            function () use ($roleId): array {
+                return [
+                    $this->cacheIdentifierGenerator->generateTag(
+                        self::ROLE_ASSIGNMENT_ROLE_LIST_IDENTIFIER,
+                        [$roleId]
+                    ),
+                    $this->cacheIdentifierGenerator->generateTag(self::ROLE_IDENTIFIER, [$roleId]),
+                ];
+            },
+            [$roleId, $offset, $limit]
+        );
+    }
+
+    public function countRoleAssignments(int $roleId): int
+    {
+        $cacheItem = $this->cache->getItem(
+            $this->cacheIdentifierGenerator->generateKey(
+                self::ROLE_ASSIGNMENTS_COUNT,
+                [$roleId],
+                true
+            )
+        );
+
+        if ($cacheItem->isHit()) {
+            $this->logger->logCacheHit(['roleId' => $roleId]);
+
+            return $cacheItem->get();
+        }
+
+        $this->logger->logCacheMiss(
+            ['roleId' => $roleId]
+        );
+
+        $userHandler = $this->persistenceHandler->userHandler();
+        $count = $userHandler->countRoleAssignments($roleId);
+
+        // Build tags for every role assignment
+        $roleAssignments = $userHandler->loadRoleAssignmentsByRoleId($roleId);
+
+        $roleAssignmentTagsClosure = $this->getRoleAssignmentTags;
+        $roleAssignmentTags = array_map(
+            static function (RoleAssignment $roleAssignment) use ($roleAssignmentTagsClosure): array {
+                return $roleAssignmentTagsClosure($roleAssignment);
+            },
+            $roleAssignments ?: []
+        );
+
+        $tags = [];
+        foreach ($roleAssignmentTags as $roleAssignmentTagArray) {
+            foreach ($roleAssignmentTagArray as $tag) {
+                $tags[] = $tag;
+            }
+        }
+
+        $tags = array_merge(
+            [
+                $this->cacheIdentifierGenerator->generateTag(
+                    self::ROLE_ASSIGNMENT_ROLE_LIST_IDENTIFIER,
+                    [$roleId]
+                ),
+                $this->cacheIdentifierGenerator->generateTag(self::ROLE_IDENTIFIER, [$roleId]),
+            ],
+            $tags
+        );
+
+        $cacheItem->set($count);
+        $cacheItem->tag($tags);
+
+        $this->cache->save($cacheItem);
+
+        return $count;
```

Comment:
This whole thing, part from https://github.com/ezsystems/ezplatform-kernel/pull/373/files#r1205653912 discussion, misses in-memory cache hits. Those dedicated methods, like `getCacheValue` are handling that. See #374 for the suggested change.